text/0052-resource-importing-support.md

```diff
@@ -0,0 +1,268 @@
+configuration (e.g. an S3 bucket having specific `bucketName` configured) but
+for some resources, the code has to rely on user input.
+
+To address this, the solution is divided into two steps:
+
+##### Physical resource identification
+
+First, identify what properties must be looked up/asked for for each resource to
+be imported. This information can be obtained by executing CloudFormation API
+action `GetTemplateSummary` with the new template. The action returns the
+property names used as "import identifiers" for each resource type used in the
+template.
+
+For each new resource (one that isn't present in the currently deployed
+template), look up the values for the "import identifiers". First, look up the
+resource (CDK construct) configuration. If not explicitly defined, interactively
+prompt user to provide the value.
+
+Based on command line flags, the compiled "resource mapping" structure is either
+passed on to the next phase, or written to the specified file.
+
+##### Import execution
+
+Based on command line flags, this phase either reads "resource mapping" from the
+specified file, or it will be passed the resource mapping directly from the
+previous phase.
+
+First, the sanity checks are performed and then the change set is created. By
+default, the change set is also executed (can be changed by passing
+`--no-execute` flag just like with regular deploy).
+
+#### Cannot import and create/change resources in the same changeset
+
+CloudFormation API doesn't allow any non-import changes in the template for
+import operation. This is a challenge for CDK, as:
+
+* `CDKMetadata` resource changes on each synthesis, so the import operation
+  never succeeds with the unmodified template. To resolve this, the proposed
+  solution restores CDKMetadata to the currently deployed version before
+  passing the template to ChangeSet creation step.
+* High-level constructs produce multiple underlying CloudFormation resources
+  and only some of those might need to be imported - the rest might need to
+  be created (e.g. `ec2.Vpc()` might require VPC being imported, but subnets
+  being created). Furthermore, CDK can't automatically determine what
+  resources to import and what to create. For this to behave correctly, the
+  solution will prompt user for each resource - whether it should be imported
+  or not. If not, the resource is removed from the template before creating
+  the change set. The remaining (not imported) resources can be created by
+  a subsequent `cdk deploy` command.
+
+#### An explicit DeletionPolicy must be set on imported resources
+
+CloudFormation import API requires all imported resources to have specified
+explicit DeletionPolicy. CDK satisfies this requirement for typical stateful
+resources (e.g. S3 buckets), but not _all_ resources (e.g. VPCs). The solution
+will inject a `DeletionPolicy` with a value of `Delete` (CloudFormation's
+default) in the template for all resources without `DeletionPolicy` set by CDK.
```

Comment:
Fuck yeah 👍

You really thought through a lot of detail here, nice job!

Reply:
thanks!